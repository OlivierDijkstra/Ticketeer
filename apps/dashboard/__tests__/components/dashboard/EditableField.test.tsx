import { TooltipProvider } from '@radix-ui/react-tooltip';
import { act, render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { afterEach, describe, expect, test, vi } from 'vitest';

import EditableField from '@/components/EditableField';

const onChange = vi.fn();

describe('EditableField', () => {
  afterEach(() => {
    onChange.mockClear();
  });

  test('it enters edit mode on button click', async () => {
    render(
      <TooltipProvider>
        <EditableField onChange={onChange} value='Test' />
      </TooltipProvider>
    );

    const editButton = screen.getByRole('button');
    await userEvent.click(editButton);

    const inputField = screen.getByRole('textbox');
    expect(inputField).toBeInTheDocument();
  });

  test('it updates the value and calls onChange', async () => {
    render(
      <TooltipProvider>
        <EditableField onChange={onChange} value='Test' />
      </TooltipProvider>
    );

    const editButton = screen.getByRole('button');
    await userEvent.click(editButton);

    const inputField = screen.getByRole('textbox');
    await userEvent.clear(inputField);
    await userEvent.type(inputField, 'Updated Value{enter}');

    expect(onChange).toHaveBeenCalledWith('Updated Value');
  });

  test('it shows error for invalid value', async () => {
    render(
      <TooltipProvider>
        <EditableField onChange={onChange} value='Test' minLength={5} />
      </TooltipProvider>
    );

    const editButton = screen.getByRole('button');
    await userEvent.click(editButton);

    const inputField = screen.getByRole('textbox');
    await userEvent.clear(inputField);
    await userEvent.type(inputField, 'abc{enter}');

    expect(
      screen.getByText('Value must be at least 5 characters long')
    ).toBeInTheDocument();
  });

  test('it shows confirmation dialog', async () => {
    render(
      <TooltipProvider>
        <EditableField onChange={onChange} value='Test' confirmation />
      </TooltipProvider>
    );

    const editButton = screen.getByRole('button');
    await userEvent.click(editButton);

    const inputField = screen.getByRole('textbox');
    await userEvent.clear(inputField);
    await userEvent.type(inputField, 'New Value{enter}');

    expect(screen.getByText('Are you absolutely sure?')).toBeInTheDocument();
  });

  test('it handles dialog actions correctly', async () => {
    render(
      <TooltipProvider>
        <EditableField onChange={onChange} value='Test' confirmation />
      </TooltipProvider>
    );

    const editButton = screen.getByRole('button');
    await userEvent.click(editButton);

    const inputField = screen.getByRole('textbox');
    await userEvent.clear(inputField);
    await userEvent.type(inputField, 'New Value{enter}');

    const continueButton = screen.getByText('Continue');
    await userEvent.click(continueButton);

    expect(onChange).toHaveBeenCalledWith('New Value');
  });

  test('it cancels edit mode on blur', async () => {
    render(
      <TooltipProvider>
        <EditableField onChange={onChange} value='Test' />
      </TooltipProvider>
    );

    const editButton = screen.getByRole('button');

    await userEvent.click(editButton);

    const inputField = screen.getByRole('textbox');

    await userEvent.clear(inputField);
    await userEvent.type(inputField, 'New Value');

    await act(async () => {
      await inputField.blur();
    });

    expect(screen.getByText('Test')).toBeInTheDocument();
    expect(onChange).not.toHaveBeenCalled();
  });

  test('it cancels edit mode on Escape key press', async () => {
    render(
      <TooltipProvider>
        <EditableField onChange={onChange} value='Test' />
      </TooltipProvider>
    );

    const editButton = screen.getByRole('button');
    await userEvent.click(editButton);

    const inputField = screen.getByRole('textbox');
    await userEvent.clear(inputField);
    await userEvent.type(inputField, 'New Value');

    await act(async () => {
      await userEvent.keyboard('{Escape}');
    });

    expect(screen.getByText('Test')).toBeInTheDocument();
    expect(onChange).not.toHaveBeenCalled();
  });

  test('it edits a textarea correctly', async () => {
    render(
      <TooltipProvider>
        <EditableField onChange={onChange} value='Test' type='textarea' />
      </TooltipProvider>
    );

    const editButton = screen.getByRole('button');
    await userEvent.click(editButton);

    const textareaField = screen.getByRole('textbox');
    await userEvent.clear(textareaField);
    await userEvent.type(textareaField, 'Updated Text{enter}');

    expect(onChange).toHaveBeenCalledWith('Updated Text');
  });

  test('it handles empty value when required', async () => {
    render(
      <TooltipProvider>
        <EditableField
          onChange={onChange}
          value='Test'
          minLength={5}
          required
        />
      </TooltipProvider>
    );

    const editButton = screen.getByRole('button');
    await userEvent.click(editButton);

    const inputField = screen.getByRole('textbox');
    await userEvent.clear(inputField);
    await userEvent.type(inputField, '{enter}');

    expect(screen.getByText('Value cannot be empty')).toBeInTheDocument();
  });

  test('it cancels changes when confirmation is not accepted', async () => {
    render(
      <TooltipProvider>
        <EditableField onChange={onChange} value='Test' confirmation />
      </TooltipProvider>
    );

    const editButton = screen.getByRole('button');
    await userEvent.click(editButton);

    const inputField = screen.getByRole('textbox');
    await userEvent.clear(inputField);
    await userEvent.type(inputField, 'New Value{enter}');

    const cancelButton = screen.getByText('Cancel');
    await userEvent.click(cancelButton);

    expect(screen.getByText('Test')).toBeInTheDocument();
    expect(onChange).not.toHaveBeenCalled();
  });
});
