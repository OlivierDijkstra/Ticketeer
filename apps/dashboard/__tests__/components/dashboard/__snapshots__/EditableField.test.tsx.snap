// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`EditableField > it renders correctly 1`] = `
<div>
  <div>
    <div
      class="mb-2 flex items-center gap-2"
    >
      <div
        class="w-full"
      >
        <span>
          Test
        </span>
      </div>
      <button
        aria-label="edit value"
        class="inline-flex items-center justify-center whitespace-nowrap text-sm font-medium transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-50 border border-input bg-background shadow-sm hover:bg-accent hover:text-accent-foreground h-9 w-9 text-primary rounded-full"
        data-state="closed"
      >
        <span
          class="sr-only"
        >
          Edit value
        </span>
        <svg
          class="lucide lucide-pencil"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M17 3a2.85 2.83 0 1 1 4 4L7.5 20.5 2 22l1.5-5.5Z"
          />
          <path
            d="m15 5 4 4"
          />
        </svg>
      </button>
    </div>
  </div>
</div>
`;
